app/src/main/java/woowacourse/shopping/di/module/Module.kt

```diff
@@ -0,0 +1,72 @@
+package woowacourse.shopping.di.module
+
+import kotlin.reflect.KClass
+import kotlin.reflect.KFunction
+import kotlin.reflect.KVisibility
+import kotlin.reflect.full.declaredMemberFunctions
+import kotlin.reflect.full.isSubclassOf
+import kotlin.reflect.full.primaryConstructor
+import kotlin.reflect.full.valueParameters
+
+abstract class Module(private val parentModule: Module? = null) {
+    protected val cache =
+        mutableMapOf<String, Any>() // 모듈은 자신의 생명이 살아있는 한, 한 번 만들었던 동일한 객체를 뱉어내야 한다.
+
+    fun <T : Any> provideInstance(clazz: Class<T>): T {
+        val injectableFunctionWithModuleMap = searchInjectableFunctions(clazz)
+
+        // 해당 객체를 만들기 위한 필요 구성 요소들을 얻는 과정.
+        return when (injectableFunctionWithModuleMap.size) {
+            0 -> createWithPrimaryConstructor(clazz) // 모듈에 정의된 메소드들 중 해당되는 것이 없다는 것은, 이 객체를 만드는데 주생성자면 충분하다는 의미.
+            else -> { // 추후 1과 else로 분리 예정
+                val injectFunction =
+                    injectableFunctionWithModuleMap.keys.first() // 메소드들 중 어노테이션으로 판단해서 일치하는 메소드를 실행시켜서 반환시킬 것임.
+                val injectModule = injectableFunctionWithModuleMap[injectFunction]
+                    ?: throw NullPointerException("${injectFunction.name} 함수가 없습니다.")
+                createWithModuleFunc(injectModule, injectFunction)
+            }
+        }
+    }
+
+    private fun <T : Any> searchInjectableFunctions(
+        clazz: Class<T>,
+    ): Map<KFunction<*>, Module> {
+        return getPublicMethodMap().filter { entry ->
+            val returnKClass = entry.key.returnType.classifier as KClass<*>
+            clazz.kotlin.isSubclassOf(returnKClass)
+        }.takeIf { it.isNotEmpty() || parentModule == null }
+            ?: parentModule?.searchInjectableFunctions(clazz) ?: mapOf()
+    }
```

Comment:
```suggestion
        return getPublicMethodMap().filter { entry ->
            val returnKClass = entry.key.returnType.classifier as KClass<*>
            clazz.kotlin.isSubclassOf(returnKClass)
        }.takeUnless { it.isEmpty() }
            ?: parentModule?.searchInjectableFunctions(clazz) ?: mapOf()
    }
```
어떠신가요? 조건문에 not을 없애고, `parentModule`이 null인 경우는 이미 뒤에서 처리하고 있으므로 고려하지 않아도 될 것 같습니다! 😄